concurrency_limiter_test.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+func makeFakeHandler(ctx context.Context, completeInFlightRequestChan chan struct{}) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		select {
+		case <-ctx.Done():
+			w.WriteHeader(http.StatusServiceUnavailable)
+		case <-completeInFlightRequestChan:
+			w.WriteHeader(http.StatusOK)
+		}
+	}
+}
+
+func doRRandRequest(ctx context.Context, t *testing.T, wg *sync.WaitGroup, cl *concurrencyLimiter) *httptest.ResponseRecorder {
```

Comment:
Is it worth returning the error so that testing.T doesn't have to be a dependency of the method?

Reply:
Minor disagree with this, because we call `doRRandRequest` about 10 times. If we added an error checking statement to each one, that'd be 20-30 lines of error checking. I would rather add a panic there, given that really shouldn't return in error. I can do either.

Reply:
That was one of the things that I was considering when asking the question. it does add some additional lines of code when the error handling is done in each test.

Reply:
I think turning it into a panic makes sense, because realistically, `req, err := http.NewRequest("GET", "/", nil)` should never fail.

Reply:
Changed this to a panic.